File: sdks/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
SENTINEL Python SDK - Basic Usage Example

This example demonstrates how to use the SENTINEL SDK to:
- Create a goal manifold
- Add goals with success criteria
- Track progress
- Validate alignment
"""

import asyncio
import os
from datetime import datetime

from sentinel_sdk import (
    SentinelClient,
    GoalStatus,
    ProbabilityDistribution,
    DistributionType,
    GoalMetadata,
)


# Configuration
SENTINEL_URL = os.environ.get("SENTINEL_URL", "http://localhost:8080")


async def main():
    """Run the example."""
    # Initialize client with context manager for automatic cleanup
    async with SentinelClient(SENTINEL_URL, timeout=30.0) as client:
        print(f"🔗 Connected to SENTINEL at {SENTINEL_URL}")
        
        # Step 1: Create a new goal manifold for a data pipeline project
        manifold = await client.create_goal_manifold(
            description="Build a real-time data processing pipeline for analytics",
            constraints=[
                "Use Python 3.11+",
                "Apache Kafka for streaming",
                "PostgreSQL for storage",
                "Docker for deployment",
                "Test coverage > 85%",
            ],
            languages=["Python", "SQL"],
            frameworks=["FastAPI", "Apache Kafka", "SQLAlchemy"],
            target_platform="cloud",
        )
        print(f"✅ Created goal manifold: {manifold.id}")
        print(f"   Root intent: {manifold.root_intent.description}")
        
        # Step 2: Add infrastructure setup goal
        setup_goal = await client.add_goal(
            manifold_id=manifold.id,
            description="Setup Docker infrastructure and Kafka cluster",
            success_criteria=[
                {"type": "file_exists", "params": {"path": "docker-compose.yml"}},
                {"type": "file_exists", "params": {"path": "Dockerfile"}},
                {"type": "api_available", "params": {"endpoint": "http://localhost:9092", "method": "GET"}},
            ],
            complexity_estimate=ProbabilityDistribution(
                mean=5.0,
                std_dev=1.0,
                min=3.0,
                max=7.0,
                distribution_type=DistributionType.NORMAL,
            ),
            value_to_root=0.15,
            status=GoalStatus.READY,
            metadata=GoalMetadata(
                tags=["infrastructure", "docker", "kafka"],
                notes="Foundation for the data pipeline",
                priority=1.0,
            ),
        )
        print(f"✅ Added infrastructure goal: {setup_goal.id}")
        
        # Step 3: Add data ingestion service goal
        ingestion_goal = await client.add_goal(
            manifold_id=manifold.id,
            description="Implement Kafka consumer for data ingestion",
            success_criteria=[
                {"type": "tests_passing", "params": {"suite": "ingestion", "min_coverage": 0.85}},
                {"type": "file_exists", "params": {"path": "src/consumers/data_consumer.py"}},
            ],
            dependencies=[setup_goal.id],
            complexity_estimate=ProbabilityDistribution(
                mean=7.0,
                std_dev=1.5,
                min=5.0,
                max=9.0,
                distribution_type=DistributionType.NORMAL,
            ),
            value_to_root=0.25,
            status=GoalStatus.PENDING,
            validation_tests=["test/consumers/test_data_consumer.py"],
            metadata=GoalMetadata(
                tags=["kafka", "consumer", "ingestion"],
                notes="Critical for data pipeline input",
                priority=0.9,
            ),
        )
        print(f"✅ Added ingestion goal: {ingestion_goal.id}")
        
        # Step 4: Add data processing service goal
        processing_goal = await client.add_goal(
            manifold_id=manifold.id,
            description="Build real-time data transformation service",
            success_criteria=[
                {"type": "tests_passing", "params": {"suite": "processing", "min_coverage": 0.9}},
                {"type": "api_available", "params": {"endpoint": "/api/v1/process", "method": "POST"}},
            ],
            dependencies=[setup_goal.id],
            complexity_estimate=ProbabilityDistribution(
                mean=8.0,
                std_dev=1.0,
                min=6.0,
                max=10.0,
                distribution_type=DistributionType.NORMAL,
            ),
            value_to_root=0.3,
            status=GoalStatus.PENDING,
            validation_tests=["test/processing/test_transformer.py"],
            metadata=GoalMetadata(
                tags=["processing", "transformation", "real-time"],
                notes="Core business logic",
                priority=1.0,
            ),
        )
        print(f"✅ Added processing goal: {processing_goal.id}")
        
        # Step 5: Add analytics API goal (depends on processing)
        api_goal = await client.add_goal(
            manifold_id=manifold.id,
            description="Create REST API for analytics queries",
            success_criteria=[
                {"type": "tests_passing", "params": {"suite": "api", "min_coverage": 0.8}},
                {"type": "api_available", "params": {"endpoint": "/api/v1/analytics", "method": "GET"}},
                {"type": "api_available", "params": {"endpoint": "/api/v1/metrics", "method": "GET"}},
            ],
            dependencies=[processing_goal.id],
            complexity_estimate=ProbabilityDistribution(
                mean=6.0,
                std_dev=1.0,
                min=4.0,
                max=8.0,
                distribution_type=DistributionType.NORMAL,
            ),
            value_to_root=0.2,
            status=GoalStatus.PENDING,
            validation_tests=["test/api/test_analytics.py"],
            metadata=GoalMetadata(
                tags=["api", "rest", "analytics"],
                notes="User-facing API",
                priority=0.8,
            ),
        )
        print(f"✅ Added API goal: {api_goal.id}")
        
        # Step 6: Mark setup goal as completed
        completed_setup = await client.update_goal_status(
            manifold_id=manifold.id,
            goal_id=setup_goal.id,
            status=GoalStatus.COMPLETED,
        )
        print(f"✅ Marked setup goal as completed")
        
        # Step 7: Check overall progress
        updated_manifold = await client.get_goal_manifold(manifold.id)
        print(f"\n📊 Project Progress:")
        print(f"   Completion: {updated_manifold.completion_percentage * 100:.1f}%")
        print(f"   Total goals: {len(updated_manifold.goals)}")
        
        # Step 8: Validate alignment
        alignment = await client.get_alignment_score(manifold.id)
        print(f"\n🎯 Alignment Score: {alignment.score:.1f}/100")
        print(f"   Confidence: {alignment.confidence * 100:.1f}%")
        
        if alignment.violations:
            print(f"   ⚠️  Violations found: {len(alignment.violations)}")
            for v in alignment.violations:
                print(f"      - {v.description} ({v.severity.value})")
        else:
            print(f"   ✅ No alignment violations")
        
        # Step 9: Validate invariants
        validation = await client.validate_invariants(manifold.id)
        print(f"\n🔒 Invariant Validation:")
        print(f"   Passed: {validation.passed}")
        if validation.violations:
            print(f"   ⚠️  Violations: {len(validation.violations)}")
        
        # Step 10: List all manifolds
        manifolds = await client.list_goal_manifolds()
        print(f"\n📋 Total manifolds in system: {len(manifolds)}")
        
        print("\n👋 Done!")


async def subscribe_example():
    """Example of subscribing to real-time updates."""
    async with SentinelClient(SENTINEL_URL) as client:
        # First create a manifold
        manifold = await client.create_goal_manifold(
            description="Test subscription",
            constraints=["Test"],
        )
        
        print(f"🔄 Subscribing to updates for manifold: {manifold.id}")
        
        # Subscribe to updates with timeout
        try:
            async for update in client.subscribe_to_updates(manifold.id):
                print(f"📡 Update: {update.type} at {update.timestamp}")
                
                # Exit after first update for demo
                break
                
        except asyncio.TimeoutError:
            print("⏱️  Subscription timed out (expected in demo)")


if __name__ == "__main__":
    # Run main example
    asyncio.run(main())
    
    # Uncomment to run subscription example
    # asyncio.run(subscribe_example())

File: sdks/typescript/examples/basic-usage.ts
/**
 * SENTINEL TypeScript SDK - Basic Usage Example
 * 
 * This example demonstrates how to use the SENTINEL SDK to:
 * - Create a goal manifold
 * - Add goals with success criteria
 * - Track progress
 * - Validate alignment
 */

import { SentinelClient, Intent, Goal, GoalStatus } from '@sentinel/sdk';

// Configuration
const SENTINEL_URL = process.env.SENTINEL_URL || 'http://localhost:8080';

async function main() {
  // Initialize client
  const client = new SentinelClient({
    baseUrl: SENTINEL_URL,
    timeout: 30000,
  });

  console.log('🔗 Connected to SENTINEL at', SENTINEL_URL);

  // Step 1: Create a new goal manifold for a web application project
  const intent: Intent = {
    description: 'Build a modern e-commerce platform with user authentication, product catalog, and payment integration',
    constraints: [
      'Use Next.js 14 with App Router',
      'PostgreSQL database with Prisma ORM',
      'Stripe for payment processing',
      'Tailwind CSS for styling',
      'Test coverage > 80%',
    ],
    languages: ['TypeScript', 'SQL'],
    frameworks: ['Next.js', 'Prisma', 'Tailwind CSS'],
    targetPlatform: 'web',
  };

  const manifold = await client.createGoalManifold(intent);
  console.log(`✅ Created goal manifold: ${manifold.id}`);
  console.log(`   Root intent: ${manifold.rootIntent.description}`);

  // Step 2: Add project setup goal
  const setupGoal = await client.addGoal(manifold.id, {
    description: 'Initialize project structure and dependencies',
    successCriteria: [
      { type: 'file_exists', params: { path: 'package.json' } },
      { type: 'file_exists', params: { path: 'tsconfig.json' } },
      { type: 'file_exists', params: { path: 'prisma/schema.prisma' } },
    ],
    dependencies: [],
    antiDependencies: [],
    complexityEstimate: {
      mean: 3.0,
      stdDev: 0.5,
      min: 2.0,
      max: 4.0,
      distributionType: 'normal',
    },
    valueToRoot: 0.1,
    status: 'ready' as GoalStatus,
    validationTests: [],
    metadata: {
      tags: ['setup', 'configuration'],
      notes: 'Foundation for the entire project',
      priority: 1.0,
    },
  });
  console.log(`✅ Added setup goal: ${setupGoal.id}`);

  // Step 3: Add authentication goal (depends on setup)
  const authGoal = await client.addGoal(manifold.id, {
    description: 'Implement JWT-based authentication system',
    successCriteria: [
      { type: 'tests_passing', params: { suite: 'auth', minCoverage: 0.85 } },
      { type: 'file_exists', params: { path: 'src/lib/auth.ts' } },
      { type: 'api_available', params: { endpoint: '/api/auth/login', method: 'POST' } },
    ],
    dependencies: [setupGoal.id],
    antiDependencies: [],
    complexityEstimate: {
      mean: 7.0,
      stdDev: 1.0,
      min: 5.0,
      max: 9.0,
      distributionType: 'normal',
    },
    valueToRoot: 0.25,
    status: 'pending' as GoalStatus,
    validationTests: ['test/auth/login.test.ts', 'test/auth/register.test.ts'],
    metadata: {
      tags: ['authentication', 'security', 'jwt'],
      notes: 'Critical security component',
      priority: 0.9,
    },
  });
  console.log(`✅ Added authentication goal: ${authGoal.id}`);

  // Step 4: Add product catalog goal
  const catalogGoal = await client.addGoal(manifold.id, {
    description: 'Build product catalog with search and filtering',
    successCriteria: [
      { type: 'tests_passing', params: { suite: 'catalog', minCoverage: 0.8 } },
      { type: 'api_available', params: { endpoint: '/api/products', method: 'GET' } },
    ],
    dependencies: [setupGoal.id],
    antiDependencies: [],
    complexityEstimate: {
      mean: 6.0,
      stdDev: 1.5,
      min: 4.0,
      max: 8.0,
      distributionType: 'normal',
    },
    valueToRoot: 0.2,
    status: 'pending' as GoalStatus,
    validationTests: ['test/catalog/products.test.ts'],
    metadata: {
      tags: ['products', 'search', 'catalog'],
      notes: 'Core e-commerce functionality',
      priority: 0.8,
    },
  });
  console.log(`✅ Added catalog goal: ${catalogGoal.id}`);

  // Step 5: Add payment integration goal (depends on auth)
  const paymentGoal = await client.addGoal(manifold.id, {
    description: 'Integrate Stripe payment processing',
    successCriteria: [
      { type: 'tests_passing', params: { suite: 'payment', minCoverage: 0.9 } },
      { type: 'api_available', params: { endpoint: '/api/payment/checkout', method: 'POST' } },
    ],
    dependencies: [authGoal.id, catalogGoal.id],
    antiDependencies: [],
    complexityEstimate: {
      mean: 8.0,
      stdDev: 1.0,
      min: 6.0,
      max: 10.0,
      distributionType: 'normal',
    },
    valueToRoot: 0.3,
    status: 'pending' as GoalStatus,
    validationTests: ['test/payment/stripe.test.ts'],
    metadata: {
      tags: ['payment', 'stripe', 'checkout'],
      notes: 'Revenue-critical feature',
      priority: 1.0,
    },
  });
  console.log(`✅ Added payment goal: ${paymentGoal.id}`);

  // Step 6: Mark setup goal as completed
  await client.updateGoalStatus(manifold.id, setupGoal.id, 'completed');
  console.log(`✅ Marked setup goal as completed`);

  // Step 7: Check overall progress
  const updatedManifold = await client.getGoalManifold(manifold.id);
  console.log(`\n📊 Project Progress:`);
  console.log(`   Completion: ${(updatedManifold.completionPercentage * 100).toFixed(1)}%`);
  console.log(`   Total goals: ${updatedManifold.goals.length}`);

  // Step 8: Validate alignment
  const alignment = await client.getAlignmentScore(manifold.id);
  console.log(`\n🎯 Alignment Score: ${alignment.score.toFixed(1)}/100`);
  console.log(`   Confidence: ${(alignment.confidence * 100).toFixed(1)}%`);
  
  if (alignment.violations.length > 0) {
    console.log(`   ⚠️  Violations found: ${alignment.violations.length}`);
    for (const v of alignment.violations) {
      console.log(`      - ${v.description} (${v.severity})`);
    }
  } else {
    console.log(`   ✅ No alignment violations`);
  }

  // Step 9: Subscribe to real-time updates
  console.log('\n🔄 Subscribing to real-time updates...');
  const unsubscribe = client.subscribeToUpdates(manifold.id, (update) => {
    console.log(`📡 Update received: ${update.type} at ${update.timestamp}`);
  });

  // Keep the subscription alive for a bit
  await new Promise((resolve) => setTimeout(resolve, 5000));
  
  // Cleanup
  unsubscribe();
  console.log('\n👋 Done!');
}

// Run the example
main().catch((error) => {
  console.error('❌ Error:', error);
  process.exit(1);
});
